addon/-private/system/class-meta.js

```diff
@@ -0,0 +1,57 @@
+import EmptyObject from 'ember-data/-private/system/empty-object';
+
+export default class ClassMeta {
+  constructor(store, modelName) {
+    this.modelName = modelName;
+    this._store = store;
+    this._modelClass = null;
+    this._idToRecord = null;
+    this._records = null;
+    this._metadata = null;
+  }
+
+  get idToRecord() {
+    return this._idToRecord || (this._idToRecord = new EmptyObject());
+  }
+
+  get records() {
+    return this._records || (this._records = []);
+  }
+
+  get metadata() {
+    return this._metadata || (this._metadata = new EmptyObject());
+  }
+
+  get modelClass() {
+    if (!this._modelClass) {
+      this._modelClass = this._store.modelFor(this.modelName);
+    }
+
+    return this._modelClass;
+  }
+
+  /* deprecated way of accessing modelClass */
+  get type() {
+    return this.modelClass;
+  }
+
+  clear() {
+    if (this._records) {
+      let records = this._records.slice();
+      let record;
+
+      for (let i = 0; i < records.length; i++) {
+        record = records[i];
+        record.unloadRecord();
+        record.destroy(); // maybe within unloadRecord
```

Comment:
ya, I think (but we should double check with someone else) we want to set `this._records.length = 0` just after  `let records = this._records.slice();`